### PR TITLE
parser: cleanup parser/scanner error.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

File: parser/scanner.l [hunk 1]
```diff
@@ -133,21 +133,26 @@ func (l *lexer) next() int {
...
+			argStr = " " + x
 	}
+
 	err := errors.New(fmt.Sprintf("line %d column %d near \"%s\"", ln, c, l.val) + argStr)
 	l.errs = append(l.errs, err)
 }
 
-func (l *lexer) err(s string, args ...interface{}) {
-	l.err0(l.line, l.col, args...)
+func (l *lexer) err(arg interface{}) {
```
Comment (siddontang) [inline parser/scanner.l]: we can support `errf(format string, args ...string)`.

btw, I think `err(args ...interface{})` may be the correct format, you can see

https://golang.org/pkg/fmt/#Sprint

Comment (qiuyesuifeng) [inline parser/scanner.l]: Actually I prefer `err(string)`, but i found that many places use `l.err(err)`, so i update function to `err(interface{})` for less modifications.
i don't choose `err(args ...interface{})` is that using `string` or `error` as input is enough now, and also if `err(arg1, arg2)`, seems have to add some separators for better print.

File: parser/scanner.l [hunk 2]
```diff
@@ -133,21 +133,26 @@ func (l *lexer) next() int {
...
-	if len(args) > 0 {
-		argStr = fmt.Sprintf(" " + args[0].(string), args[1:]...)
+	switch x := arg.(type) {
+		case error:
+			argStr = " " + x.Error()
+		case string:
+			argStr = " " + x
 	}
+
 	err := errors.New(fmt.Sprintf("line %d column %d near \"%s\"", ln, c, l.val) + argStr)
```
Comment (siddontang) [inline parser/scanner.l]: use fmt.Errorf instead

File: parser/scanner.l [hunk 3]
```diff
@@ -133,21 +133,26 @@ func (l *lexer) next() int {
 	return l.c
 }
 
-func (l *lexer) err0(ln, c int, args ...interface{}) {
+func (l *lexer) err0(ln, c int, arg interface{}) {
 	var argStr string
-	if len(args) > 0 {
-		argStr = fmt.Sprintf(" " + args[0].(string), args[1:]...)
+	switch x := arg.(type) {
```
Comment (siddontang) [inline parser/scanner.l]: you can use %v directly in following fmt format. no need to check type here.

File: parser/parser.y [hunk 4]
```diff
@@ -889,12 +889,12 @@ CreateIndexStmt:
 	{
 		indexName, tableIdent, colNameList := $4.(string), $6.(table.Ident), $8.([]*coldef.IndexColName)
 		if strings.EqualFold(indexName, tableIdent.Name.O) {
-			yylex.(*lexer).err("", "index name collision: %s", indexName)
+			yylex.(*lexer).err(fmt.Sprintf("index name collision: %s", indexName))
```
Comment (siddontang) [inline parser/parser.y]: supporting errf may be more convenient than using fmt.Sprintf here.

Comment (qiuyesuifeng) [inline parser/parser.y]: Agreed, add a `errf` function is much better.

Comment (coocood) [inline parser/parser.y]: I checked out the `err` method，the first string argument is discarded.
```go
func (l *lexer) err(s string, args ...interface{}) {
	l.err0(l.line, l.col, args...)
}
```

Comment (coocood) [inline parser/parser.y]: Ah, you fixed it in this PR.

**Comment by qiuyesuifeng** [comment]
@zimulala  PTAL

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**